### PR TITLE
cdo with cmor support

### PR DESCRIPTION
**Pull request by wachsylon**
This changes enable the integration of cmor into the cdos. Therefore, the ossp-uuid needs to be linked from CDIs. That is the main difference.

### CODE REVIEW

**Comment by wachsylon** [comment]
OK could it be that I have to specify the addition chanell somewhere?
This would be -c pcmdi .

**Comment by wachsylon** [comment]
Ok CMOR is in the pcmdi channel https://anaconda.org/PCMDI but the conda-forge recipes doc says: "Avoid Dependencies Outside of Conda-Forge". So is it not possible to use this channel?

**Comment by ocefpaf** [comment]
You will need to add the PCMDI to [staged-recipes](https://github.com/conda-forge/staged-recipes#getting-started).

**Comment by wachsylon** [comment]
It would be enough to have a staged-recipe for CMOR, correct? Since I will (and probably have to) ask the CMOR developers for permissions: What is the advantage of having a conda-forge channel when also having an anaconda channel? Would someone then need to keep an eye on that CMOR feedstock and update it from time to time? How independent is the administration of that feedstock? Is it more like a copy of that anaconda channel? Thanks for your fast response

**Comment by wachsylon** [comment]
If I install CMOR via conda with `-c conda-forge`, only a few packages are listed without a channel specified - this is the default channel, correct? 
intel-openmp
libgcc
libgfortran
libstdcxx-ng
mkl
numpy
The rest is from conda-forge. CDO's also require some packages of those. Therefore, only CMOR would be additionally needed.

@dnadeau4 Hi Denis, I am trying to enable the installation of CDO with CMOR support via Conda. Since CDO is in the conda-forge channel, the installation request CMOR to be in that channel too. Would it be OK for you to have a staged-recipe for CMOR in the conda-forge channel? The last commit of @ocefpaf gives a detailed overview of what is happening with the resulting feedstock.

**Comment by ocefpaf** [comment]
> Therefore, only CMOR would be additionally needed.

:+1:

**Comment by dnadeau4** [comment]
@taylor13  @durack1  Is it `ok` to add `CMOR`to `conda-forge` for `CDO`?

**Comment by durack1** [comment]
@wachsylon @ocefpaf two questions.

1. What is the problem with using:
`conda install -c conda-forge -c pcmdi cdo cmor`

And 
2. additionally, we've been having problems with the `conda-forge` channel netcdf-fortran libraries in particular, with these often being relabeled under the [`broken` label](https://anaconda.org/conda-forge/netcdf-fortran) and making it impossible for us to reuse [the *.yml files that we've been saving to reproduce frozen and released versions](https://github.com/CDAT/cdat/tree/master/conda)

@doutriaux1 @taylor13 @gleckler1

**Comment by ocefpaf** [comment]
> What is the problem with using

Please see https://github.com/conda-forge/cdo-feedstock/pull/38#issuecomment-381981758

> additionally, we've been having problems with the conda-forge channel netcdf-fortran libraries 

Please report that issue in https://github.com/conda-forge/netcdf-fortran-feedstock/issues

Usually a library is posted in the broken channel when the dependencies are no longer consistent with the stack. If the new build is not OK for your stack you should let us know so we can arrange new builds.

**Comment by wachsylon** [comment]
@durack1 @dnadeau4 @taylor13
> What is the problem with using:
>` conda install -c conda-forge -c pcmdi cdo cmor`

Tests are failing of building CDO with CMOR because CMOR is not in the conda-forge channel. Other packages which try to link CMOR will have similar problems if they are in a conda-forge feedstock.

The other way around, I would like to know why you have your own PCMDI branch.
@ocefpaf says
> Packages built in a different tools stack will often be incompatible with another channel

If I try to install CMOR without` -c conda-forge`, it will not work because at least two packages are only in the conda-forge channel. What do you think about the statement of @ocefpaf ?

Despite from that, it seems to me that the additional work for having a conda-forge feedstock in addition to the one in the PCMDI channel is limited since bots will do most of the work. I would be very glad to be able to offer a simple way to install CDO with CMOR via conda.

**Comment by ocefpaf** [comment]
@wachsylon note that the conda-forge recipes are BSD 3-Clause, so you have an alternative here.
You could pull the `cdo` recipe, and its dependencies, into the `PCMDI` channel.

I, of course, will argue that the other way around is better for the greater community that adopted conda-forge, but we are 100% OK with people using our recipes in their own third party channels.

**Comment by wachsylon** [comment]
I will clarify if that is an option. Thanks

**Comment by dnadeau4** [comment]
@wachsylon I got the `ok` to install `CMOR` into conda-forge.  

We will have to make clear to users that issues created for `CMOR` are not `CDO` issues.  

This means that people will have to create a `native` `CMOR` script if they want me to help them with the issue even when using `CDO`.  Otherwise, I will forward the issue to `CDO`.

**Comment by dnadeau4** [comment]
I will start the staged-recipe for  `CMOR`  next week.

**Comment by wachsylon** [comment]
Great! Thank you very much.

> We will have to make clear to users that issues created for CMOR are not CDO issues.
> 
> This means that people will have to create a native CMOR script if they want me to help them with the issue even when using CDO.
My suggestion is: If `cdo cmor` fails, I will guide the user to the CDO support or directly me. If the problem is a `CMOR` issue, I will contact you afterwards and attach a native script if necessary.

> Otherwise, I will forward the issue to CDO.
Yes! I will try to prevent users from contacting you if `cdo cmor` fails . If they do it anyway, do not hesitate to foward the issue.

**Comment by wachsylon** [comment]
@dnadeau4 
Did you make progress with the staged recipe?

**Comment by dnadeau4** [comment]
@wachsylon not really... I have 3 jobs these days and some critical issues to solve.  @doutriaux1 might be able to help.  I used to work on CMOR 1 day/week.

**Comment by doutriaux1** [comment]
@wachsylon as per my other email, I took this over, will work on it next.

**Comment by wachsylon** [comment]
Regarding to circleci:
When I call `conda-build` on my computer I get:

`checking for library containing uuid_create... -L/home/wachsmann/miniconda3/conda-bld/cdo_1530018563033/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pla/lib -lossp-uuid `

So I am not sure what goes wrong...

**Comment by wachsylon** [comment]
@ocefpaf Can I access on the config.log file from the circleci? Do you have a hint what I can do next?

File: recipe/meta.yaml [hunk 1]
```diff
@@ -6,7 +6,7 @@ package:
 
 source:
   url: https://code.zmaw.de/attachments/download/17451/cdo-{{ version }}.tar.gz
-  sha256: cc6993e96a20d4b0829cb8f67dbc798e2ee7fa734198054e6c85bcc7d86e3f2e
+  sha256: 906e03f5518488726f8447c6ff4f7fa8
```
Comment (ocefpaf) [inline recipe/meta.yaml]: sha256 is probably not that short.
Take a look at

https://github.com/conda-forge/staged-recipes#2-how-do-i-populate-the-hash-field

for some tips on how to calculate the `sha256` when that is not available upstream.

**Comment by ocefpaf** [comment]
> @ocefpaf Can I access on the config.log file from the circleci?

You can ssh into CircleCI but I never really do that.
I find it easier to run the Docker image locally.

> Do you have a hint what I can do next?

You can,

```
docker pull condaforge/linux-anvil
```

to get the image, and then

```
docker run --net=host -it --rm -v $(pwd):/home/ condaforge/linux-anvil /bin/sh
```

to "enter" the image and mount the current directory to "home" (so issue that command from the root directory).

Then you can fire up `conda build recipe` and debug it.

**Comment by wachsylon** [comment]
Oh no, I always thought (because of the name) that package util-linux contains more than only libuuid. However, it is strange that the ossuuid package names its lib the same. All the time I was wondering where the hell is uuid_create..
Now it could work.

**Comment by wachsylon** [comment]
Hi @ocefpaf @Try2Code,
I am confused, I actually did not change the configure call for Darwin (I just tried the openmp option from the other PR) but get a complete different Travis CI test with failing tests..
Do you have an idea what is happening? And is there a osx image for Docker or what is the usual way to reproduce the Travis tests? Using a VM?
Best regards and thank you very much for helping
Fabi

**Comment by ocefpaf** [comment]
> Do you have an idea what is happening? And is there a osx image for Docker or what is the usual way to reproduce the Travis tests? Using a VM?

Sorry. I don't know the first thing about OS X and the lack of images make it hard for us to debug this. Your best bet is to find a Mac person.

@isuruf do you see anything obvious that we are missing?

**Comment by wachsylon** [comment]
Nevertheless, thank you for the answer.
I found a Mac! I will try it. And fix it.

**Comment by isuruf** [comment]
This recipe is using g++, gcc on Mac which is not a supported configuration in conda-forge.

**Comment by isuruf** [comment]
@conda-forge-admin, please rerender

**Comment by wachsylon** [comment]
@isuruf Thank you very much for helping!
@ocefpaf Could we now make a 

> dev release in a sub-channel

and how would it work?

**Comment by wachsylon** [comment]
Yes, that is perfect for me.

**Comment by ocefpaf** [comment]
B/c this one is pointing to `master` I cannot merge it, but I'll pick the new commit and put it in #41 so we can get this released ASAP.

**Comment by isuruf** [comment]
FYI, if you click edit on a PR, you can change the target branch from master to dev

**Comment by ocefpaf** [comment]
Oh, GitHub restored that? I used to do this all the time and I recall that they removed it some time ago. Good to know! Thanks!!

**Pull request closed without merging**